File: src/free_proxy/services/auto_switch.py
from __future__ import annotations

from free_proxy.config import Settings
from free_proxy.domain.enums import NodeStatus, ProxyPolicyMode
from free_proxy.domain.models import ProxyNodeRead, TunnelStartResult
from free_proxy.infrastructure.database.repositories import (
    ProxyNodeRepository,
    SettingsRepository,
)
from free_proxy.services.gateway import GatewayService
from free_proxy.services.proxy_pool import ProxyPoolService


class AutoSwitchService:
    def __init__(
        self,
        app_settings: Settings,
        nodes: ProxyNodeRepository,
        settings: SettingsRepository,
        pool: ProxyPoolService,
        gateway: GatewayService,
    ) -> None:
        self._app_settings = app_settings
        self._nodes = nodes
        self._settings = settings
        self._pool = pool
        self._gateway = gateway

    async def switch(self) -> TunnelStartResult | None:
        settings = await self._settings.get()
        if not settings.connection_enabled:
            return None
        if settings.routing_mode is ProxyPolicyMode.FIXED:
            if not settings.fixed_node_id:
                return None
            return await self._gateway.activate(settings.fixed_node_id)

        excluded: set[str] = set()
        active_id = self._gateway.status().active_node_id
        if active_id:
            excluded.add(active_id)
        for _ in range(3):
            candidate = await self._select_excluding(excluded)
            if candidate is None:
                await self._gateway.disconnect_only()
                return None
            result = await self._gateway.activate(candidate.id)
            if result.success:
                return result
            excluded.add(candidate.id)
            await self._nodes.blacklist(
                candidate.id,
                result.message,
                self._app_settings.invalid_backoff_seconds,
            )
        return None

    async def handle_unexpected_exit(self) -> None:
        settings = await self._settings.get()
        if not settings.connection_enabled:
            return
        if settings.routing_mode is ProxyPolicyMode.FIXED and settings.fixed_node_id:
            await self._gateway.activate(settings.fixed_node_id)
        else:
            await self.switch()

    async def _select_excluding(self, excluded: set[str]) -> ProxyNodeRead | None:
        try:
            candidates = await self._nodes.list_nodes(
                limit=1000,
                status=NodeStatus.READY,
                current_only=True,
            )
        except TypeError:
            candidates = await self._nodes.list_nodes(limit=1000, status=NodeStatus.READY)
        settings = await self._settings.get()
        candidates = self._pool.apply_filters(candidates, settings)
        candidates = [node for node in candidates if node.id not in excluded]
        candidates.sort(key=lambda node: self._pool.sort_key_for(node, settings))
        return candidates[0] if candidates else None

File: internal/services/autoswitch.go
package services

import (
	"context"

	"github.com/masteralanlab/free-proxy/internal/config"
	"github.com/masteralanlab/free-proxy/internal/domain"
	"github.com/masteralanlab/free-proxy/internal/store"
)

// AutoSwitchService picks a replacement exit after a failure, blacklisting nodes
// that fail to activate.
type AutoSwitchService struct {
	cfg          *config.Config
	nodes        *store.NodeRepository
	settingsRepo *store.SettingsRepository
	pool         *ProxyPoolService
	gateway      *GatewayService
}

// NewAutoSwitchService constructs an AutoSwitchService.
func NewAutoSwitchService(cfg *config.Config, nodes *store.NodeRepository, settingsRepo *store.SettingsRepository, pool *ProxyPoolService, gateway *GatewayService) *AutoSwitchService {
	return &AutoSwitchService{cfg: cfg, nodes: nodes, settingsRepo: settingsRepo, pool: pool, gateway: gateway}
}

// Switch activates the best alternate node, or the fixed node in fixed mode.
func (s *AutoSwitchService) Switch(ctx context.Context) (*domain.TunnelStartResult, error) {
	settings, err := s.settingsRepo.Get(ctx)
	if err != nil {
		return nil, err
	}
	if !settings.ConnectionEnabled {
		return nil, nil
	}
	if settings.RoutingMode == domain.PolicyFixed {
		if settings.FixedNodeID == nil || *settings.FixedNodeID == "" {
			return nil, nil
		}
		res, err := s.gateway.Activate(ctx, *settings.FixedNodeID)
		return &res, err
	}

	excluded := map[string]bool{}
	if active := s.gateway.Status().ActiveNodeID; active != nil {
		excluded[*active] = true
	}
	for i := 0; i < 3; i++ {
		candidate, err := s.selectExcluding(ctx, excluded)
		if err != nil {
			return nil, err
		}
		if candidate == nil {
			s.gateway.DisconnectOnly(ctx)
			return nil, nil
		}
		res, err := s.gateway.Activate(ctx, candidate.ID)
		if err == nil && res.Success {
			return &res, nil
		}
		excluded[candidate.ID] = true
		msg := "activation failed"
		if res.Message != "" {
			msg = res.Message
		} else if err != nil {
			msg = err.Error()
		}
		_ = s.nodes.Blacklist(ctx, candidate.ID, msg, s.cfg.InvalidBackoff())
	}
	return nil, nil
}

// HandleUnexpectedExit reconnects after an unexpected tunnel exit.
func (s *AutoSwitchService) HandleUnexpectedExit(ctx context.Context) {
	settings, err := s.settingsRepo.Get(ctx)
	if err != nil || !settings.ConnectionEnabled {
		return
	}
	if settings.RoutingMode == domain.PolicyFixed && settings.FixedNodeID != nil && *settings.FixedNodeID != "" {
		_, _ = s.gateway.Activate(ctx, *settings.FixedNodeID)
		return
	}
	_, _ = s.Switch(ctx)
}

func (s *AutoSwitchService) selectExcluding(ctx context.Context, excluded map[string]bool) (*domain.ProxyNodeRead, error) {
	candidates, err := s.nodes.ListNodes(ctx, store.NodeFilter{Status: string(domain.NodeReady), CurrentOnly: true}, 1000, 0)
	if err != nil {
		return nil, err
	}
	settings, err := s.settingsRepo.Get(ctx)
	if err != nil {
		return nil, err
	}
	candidates = ApplyFilters(candidates, settings, false)
	filtered := candidates[:0]
	for _, n := range candidates {
		if !excluded[n.ID] {
			filtered = append(filtered, n)
		}
	}
	SortCandidates(filtered, settings)
	if len(filtered) == 0 {
		return nil, nil
	}
	best := filtered[0]
	return &best, nil
}
